### PR TITLE
Adding Notifications to CosmosResponseProcessor

### PR DESCRIPTION
**Pull request by staladay**
## Description
Moving the GetInnerMostException extension method to the Fhir.Core extensions rather then having the Fhir.CosmosDb library take a whole project dependency just for an extension method and to group with other existing exception extension methods.

I also added emiting ExceptionNotifications to the CosmosResponseProcessor since most, if not all, of these exceptions get handled before the ExceptionNotificationMiddleware gets called and I need to be notified of these exceptions down stream.

## Related issues
Addresses [issue #].

## Testing
Ran all related tests locally and will do some other tests in the managed service.

## FHIR Team Checklist
- [ ] **Update the title** of the PR to be succinct and less than 50 characters
- [ ] **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- [ ] Tag the PR with the type of update: **Bug**, **Dependencies**, **Enhancement**, or **New-Feature**
- [ ] Tag the PR with **Azure API for FHIR** if this will release to the managed service
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Core/Extensions/ExceptionExtensions.cs [hunk 1]
```diff
@@ -39,5 +40,19 @@ public static bool IsRequestEntityTooLarge(this Exception e)
         {
             return e is RequestEntityTooLargeException || e?.InnerException is RequestEntityTooLargeException;
         }
+
+        public static Exception GetInnerMostException(this Exception input)
```
Comment (Ivanidzo4ka) [inline src/Microsoft.Health.Fhir.Core/Extensions/ExceptionExtensions.cs:44]: >GetInnerMostException [](http://example.com/codeflow?start=32&length=21)

[https://docs.microsoft.com/en-us/dotnet/api/system.exception.getbaseexception?view=net-5.0](https://docs.microsoft.com/en-us/dotnet/api/system.exception.getbaseexception?view=net-5.0)

would it work for you?

[https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Private.CoreLib/src/System/Exception.cs#L65](https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Private.CoreLib/src/System/Exception.cs#L65)

works a bit differently for aggregate exception, tho

[https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Private.CoreLib/src/System/AggregateException.cs#L225](https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Private.CoreLib/src/System/AggregateException.cs#L225)

File: src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CosmosResponseProcessor.cs [hunk 2]
```diff
@@ -47,41 +49,47 @@ public CosmosResponseProcessor(RequestContextAccessor<IFhirRequestContext> fhirR
...
-        public void ProcessErrorResponse(HttpStatusCode statusCode, Headers headers, string errorMessage)
+        public async Task ProcessErrorResponse(HttpStatusCode statusCode, Headers headers, string errorMessage)
         {
             if (statusCode == HttpStatusCode.TooManyRequests)
             {
                 string retryHeader = headers["x-ms-retry-after-ms"];
-                throw new RequestRateExceededException(int.TryParse(retryHeader, out int milliseconds) ? TimeSpan.FromMilliseconds(milliseconds) : null);
+                var exception = new RequestRateExceededException(int.TryParse(retryHeader, out int milliseconds) ? TimeSpan.FromMilliseconds(milliseconds) : null);
+                await EmitExceptionNotificationAsync(statusCode, exception);
+                throw exception;
```
Comment (Ivanidzo4ka) [inline src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CosmosResponseProcessor.cs:67]: nit:
I don't like code duplication, so I would do following:

```
Exception exception = null;

if (statusCode == HttpStatusCode.TooManyRequests)
{
 var exception = new RequestRateExceededException(int.TryParse(retryHeader, out int milliseconds) ? TimeSpan.FromMilliseconds(milliseconds) : null);
}
else if () {
...
}

if (exception!=null){
 await EmitExceptionNotificationAsync(statusCode, exception);
  throw exception;
}

**Comment by Ivanidzo4ka** [comment]
Code looks good for me, feel free to move it from draft to active PR and I'll be happy to approve it.

**Comment by CaitlinV39** [comment]
@staladay - Are you ready for this PR to be reviewed for merging?

**Comment by staladay** [comment]
@CaitlinV39 No. I have been unable to test this change in PaaS. I am leaving it as a draft until I can test it in TIP.

File: src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CosmosResponseProcessor.cs (hunk 2 above)
Comment (staladay) [reply src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CosmosResponseProcessor.cs]: Done

File: src/Microsoft.Health.Fhir.Core/Extensions/ExceptionExtensions.cs (hunk 1 above)
Comment (staladay) [reply src/Microsoft.Health.Fhir.Core/Extensions/ExceptionExtensions.cs]: System.GetBaseException() works for my purposes.

**Comment by staladay** [comment]
@Ivanidzo4ka @CaitlinV39 I was able to finish testing these changes last Thursday and they are ready to officially be reviewed.

**Review by Ivanidzo4ka — APPROVED** [review]
:shipit:

**Pull request merged**